[PR] Feature/security sh

bbubbush opened the pull request.

bbubbush committed:
fix: Controller Naming role 적용

bbubbush committed:
fix: VO, DTO 사용 및 생성 규칙에 맞게 변경

bbubbush committed:
fix: 모든 Repository를 JpaRepository로 통일되도록 수정

bbubbush committed:
fix: @DynamicUpdate 추가 및 @JsonIgnore 제거

bbubbush committed:
fix: 회원등록할 때 관심항목이 없다면 생성하도록 변경

bbubbush committed:
fix: 코드컨벤션에 맞춰 stream 변경

bbubbush committed:
fix: 엔티티에 생성일자, 변경일자 컬럼 자동으로 등록되도록 수정

bbubbush committed:
fix: 회원정보 변경할 때, 회원관심주제 전체 삭제 후 다시 등록하도록 수정

mike6321 committed:
[junwoo.choi] fix : 코드리뷰 반영

bbubbush committed:
Merge pull request #10 from mike6321/hotfix/member

fix: 회원코드 컨벤션에 맞춰 수정

mike6321 committed:
[junwoo.choi] fix : test code

mike6321 committed:
Merge branch 'hotfix/mission' into dev

bbubbush committed:
fix: when -> given으로 TDD에 맞게 다시 변경

mike6321 committed:
Merge remote-tracking branch 'origin/hotfix/member' into dev

bbubbush committed:
Feat: 시큐리티 이메일 로그인 기능 추가

bbubbush committed:
Feat: 시큐리티 인증/인가 필터 분리

bbubbush committed:
Delete: 사용하지 않는 클래스 삭제

mike6321 commented on src/main/java/com/mission/config/SecurityConfig.java:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  @Bean
+  DaoAuthenticationProvider authenticationProvider(){
```
혹시 DaoAuthenticationProvider를 Bean으로 새로 등록하신 이유가 있을까요~?

bbubbush replied:
이래저래 붙여보면서 디버깅 하다 그대로올라갔네요  제거할게요 없어도 동작에 큰 의미가 없습니다 ㅎㅎ

mike6321 commented on src/main/java/com/mission/domain/Grade.java:
```diff
@@ -1,15 +1,17 @@
 @AllArgsConstructor(access = AccessLevel.PRIVATE)
-@Builder
-public class Grade {
+@DynamicUpdate @Builder
```
오 저도 @DynamicUpdate 적용해야할 것 같습니다!👍🏻

mike6321 commented on src/main/java/com/mission/domain/MemberOfTopicInterest.java:
```diff
@@ -10,18 +10,16 @@
     @Column(name = "member_of_topic_of_interest_id")
     private Long memberOfTopicOfInterestId;
-    @ManyToOne(fetch = FetchType.LAZY)
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
CasecadeType을 PERSIST로 선언한 이유가 있으실까요~?

bbubbush replied:
저도 이유를 모르겠네요 ㅋㅋ  BaseTimeEntity를 적용하는 코드에서 ALL -> PERSIST로 변경되었는데 딱히 그렇게 될 이유가 없었는데 ㅎㅎ  다시 돌려놓도록 하겠습니다

mike6321 commented on src/main/java/com/mission/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,71 @@
+import java.time.LocalDateTime;
+import java.util.stream.Collectors;
+
+public class JwtAuthenticationFilter extends UsernamePasswordAuthenticationFilter {
```
오 저도 AuthenticationFilter 네이밍으로 가져가야할 것 같습니다~!
형은 UsernamePasswordAuthenticationFilter 상속에 대해서 어떻게 생각하시는지요~?

bbubbush replied:
저는 이메일 / 패스워드  로그인도 구조적으로는 Username과 Password로 분류된다고 생각해서 이걸 사용하는거에 대해 상관 없다 는 생각입니다

다만 UsernamePasswordAuthenticationFilter의 구조를 이해하기 위해 AbstractAuthenticationProcessingFilter를 구현하는 것이 좋은 경험이라고 생각합니다

mike6321 replied:
넵 이해완료하였습니다! 코멘트 감사합니다 👍

mike6321 commented on src/main/java/com/mission/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,71 @@
+public class JwtAuthenticationFilter extends UsernamePasswordAuthenticationFilter {
+
+  private final AuthenticationManager authenticationManager;
+  private static final AntPathRequestMatcher ANT_PATH_REQUEST_MATCHER = new AntPathRequestMatcher("/api/auth/**", "POST");
```
로그인 API URL 룰을 어떻게 맞출까요~?
저는 /login 로 하긴 했습니다!

bbubbush replied:
이건 통일성이 필요해보이는데 로그인방법이 늘어날 여지가 있는만큼 1depth로 표현하는건 무리가 있어보이니 2depth 구조로 가는게 어떨까요?
예를들면 /login/email,   /login/nickname  이런식이요

mike6321 commented on src/main/java/com/mission/service/SecurityUserDetailService.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  private UserDetails createUser(Member member) {
+    return User.builder()
```
code convention 08

mike6321 commented on src/main/java/com/mission/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,86 @@
+  @Override
+  protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
+    final String token = getJwtTokenFromRequestHeader(request);
+
```
code convention 02

mike6321 commented on src/main/java/com/mission/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,86 @@
+      chain.doFilter(request, response);
+      return;
+    }
+
```
code convention 02

mike6321 commented on src/main/java/com/mission/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,86 @@
+
+  private Authentication getAuthentication(String token) {
+    final Jws<Claims> claims = getClaims(token);
+
```
code convention 02

mike6321 commented on src/main/java/com/mission/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,86 @@
+public class JwtAuthorizationFilter extends BasicAuthenticationFilter {
+
+  public JwtAuthorizationFilter(AuthenticationManager authenticationManager) {
+    super(authenticationManager);
```
SecurityConfig에서 secret key를 주입해주는 형태는 어떠실까요?

mike6321 commented on src/main/java/com/mission/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,86 @@
+      claims = Jwts.parser()
+        .setSigningKey(JwtProperties.SECRET_KEY.toString())
+        .parseClaimsJws(token.replace(JwtProperties.TOKEN_PREFIX.toString(), ""));
+    } catch (ExpiredJwtException e) {
```
추가적인 catch 옵션을 설정해주면 오류를 판단하는데 유용할 것 같습니다!

mike6321 commented on src/main/java/com/mission/security/JwtProperties.java:
```diff
@@ -0,0 +1,25 @@
+
+public enum JwtProperties {
+
+  SECRET_KEY("12345678901234567890123456789012")
```
secret key를 yml에 정의하는게 어떨까요~?

mike6321 replied:
추가적으로 Base64로 암호화된 값을 넣는것이 좋아보입니다!

mike6321 commented on src/main/java/com/mission/config/SecurityConfig.java:
```diff
@@ -0,0 +1,63 @@
+        .formLogin().disable()
+        .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+      .and()
+        .addFilter(new JwtAuthenticationFilter(authenticationManager()))
```
authenticationManager()를 따로 구현하신 이유가 궁금합니다!

mike6321 commented on src/main/java/com/mission/controller/LoginController.java:
```diff
@@ -0,0 +1,24 @@
+  public ResponseEntity<String> loginByEmail(@RequestBody ReqLoginByEmail reqLoginByEmail) {
+    log.info("email :: {}", reqLoginByEmail.getEmail());
+    log.info("password :: {}", reqLoginByEmail.getPassword());
+    return ResponseEntity.ok("");
```
Response를 무엇을 반환해야할지 논의가 필요해보이네요~! 저도 따로 정의하진 않았습니다!

mike6321 commented on src/main/java/com/mission/dto/member/ReqLoginByEmail.java:
```diff
@@ -0,0 +1,16 @@
+  private String email;
+  private String password;
+
+
```
code convention 01

mike6321 commented on src/main/java/com/mission/dto/member/ReqLoginByEmail.java:
```diff
@@ -0,0 +1,16 @@
+@AllArgsConstructor
+public class ReqLoginByEmail {
+
+  private String email;
```
4칸 개행

mike6321 commented on src/main/java/com/mission/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,71 @@
+      e.printStackTrace();
+    }
+
+    assert reqLoginByEmail != null;
```
프로덕션 코드에서 단언문의 사용이 특정 경우에서 장애를 일으킨다고 알고 있는데
확인이 필요해 보입니다

mike6321 commented on src/main/java/com/mission/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,71 @@
+        .collect(Collectors.joining()))
+      .setSubject(authResult.getName())
+      .setExpiration(Timestamp.valueOf(LocalDateTime.now().plusMinutes(30L)))
+      .signWith(SignatureAlgorithm.HS512, JwtProperties.SECRET_KEY.getValue())
```
Secret key를 12345678901234567890123456789012 로 선언하셨는데 
Base64로 암호화된 값을 넣는게 좋지 않을까요~?

mike6321 commented on src/main/java/com/mission/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  private List<SimpleGrantedAuthority> getAuthorization(Jws<Claims> claimsJws) {
+    final String authorizationText = (String) claimsJws.getBody().get(JwtProperties.AUTHORIZATION_KEY.toString());
```
파라미터에 final을 붙인다면 메서드 내부에서 final을 선언을 안해도되지 않을까요~?
ex (final Jws<Claims> clainsJws)

mike6321 commented on src/main/java/com/mission/domain/Member.java:
```diff
@@ -30,24 +32,31 @@ public class Member {
-          .build();
+    public static Member createMember(ReqCreateMember memberCreateVO, List<MemberOfTopicInterest> topicOfInterests, PasswordEncoder passwordEncoder) {
+    final Member createMember =
+        Member.builder()
```
code convention 08

mike6321 commented on src/main/java/com/mission/config/SecurityConfig.java:
```diff
@@ -0,0 +1,63 @@
+@RequiredArgsConstructor
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+  private final SecurityUserDetailService userDetailService;
```
provider를 구현하지 않는다면 SecurityUserDetailService는 주입받을 필요가 없어보네요!!

bbubbush replied:
맞네요 ㅎㅎ 굳이 주입해줄 필요 없으니 삭제하도록 할게요

mike6321 commented on src/main/java/com/mission/config/SecurityConfig.java:
```diff
@@ -0,0 +1,63 @@
+  @Override
+  protected void configure(HttpSecurity http) throws Exception {
+    http
+        .httpBasic().disable()
```
저도 해당 부분 적용해도록 하겠습니다!

bbubbush replied:
어떤걸 적용해보시는거죠? 코드리뷰 위치로는 확인이 잘 안되네요 ㅜㅜㅋㅋ

mike6321 replied:
아하 제가 httpBasic을 disable 처리를 안했습니다!
![image](https://user-images.githubusercontent.com/33277588/172389375-20d37fd7-e706-4af3-9c39-bca7f8d52783.png)

mike6321 commented on src/main/java/com/mission/service/MemberService.java:
```diff
@@ -1,89 +1,116 @@
+
+    memberOfTopicOfInterestRepository.deleteByMember(findMember);
+    findMember.deleteTopicOfInterests();
+
```
code convention 02

bbubbush replied:
코드 컨벤션 1번이 클래스 위 아래 공백 한줄 맞죠??

잘 적용된거 같은데 확인 플리즈

mike6321 replied:
잘 못 말씀드렸네요! 
메서드 라인공백 x  code convention 02 번 입니다 👍

mike6321 commented on src/main/java/com/mission/service/SecurityUserDetailService.java:
```diff
@@ -0,0 +1,35 @@
+@RequiredArgsConstructor
+@Transactional
+public class SecurityUserDetailService implements UserDetailsService {
+
```
code convention 01

bbubbush replied:
코드 컨벤션 1번이 클래스 위 아래 공백 한줄 맞죠??

이것도 잘 적용된거 같은데 확인 플리즈!

mike6321 replied:
앗 제가 잘못 보았네요! 죄송합니다 💯

mike6321 commented on src/main/java/com/mission/service/SecurityUserDetailService.java:
```diff
@@ -0,0 +1,35 @@
+public class SecurityUserDetailService implements UserDetailsService {
+
+  private final MemberRepository memberRepository;
+
```
code convention 01

bbubbush replied:
코드 컨벤션 1번이 클래스 위 아래 공백 한줄 맞죠??

한번 더 확인 플리즈플리즈 : )

mike6321 replied:
이것 또한 동일합니다! 죄송합니다 💯

mike6321 reviewed (commented):
[PR] Feature/security sh review 01

bbubbush commented:
1차 코드리뷰내용 피드백 드립니다 ㅎㅎ 감사합니다

bbubbush committed:
Refactor: 1차 코드리뷰 내용에 따라 불필요한 코드 삭제

- Jwts<Clames> 사용하는 부분을 Clames을 사용하도록 변경
- Cascade.PERSIST -> Cascade.ALL 롤백
- SecurityConfig 내 불필요한 구현체 삭제

mike6321 commented on src/main/java/com/mission/domain/Member.java:
```diff
@@ -30,24 +32,31 @@ public class Member {
     private Grade grade;
+    @Column(name = "password")
+    private String password;
+//    @ElementCollection(fetch = FetchType.EAGER)
```
주석된 코드 삭제해보심이 어떠신지요~?⭐️

mike6321 commented on src/test/java/com/mission/controller/MemberControllerTest.java:
```diff
@@ -33,15 +35,16 @@
 @SpringBootTest
 @AutoConfigureMockMvc
```
해당 어노테이션의 역할에 대하서 한번 가르침을 받고자 합니다!

mike6321 reviewed (commented):
[PR] Feature/security sh 2차 코드리뷰
